substrate-archive/src/database/queries.rs

```diff
@@ -194,6 +221,57 @@ pub(crate) fn blocks_paginated<'a>(
 	})
 }
 
+/// Get up to `max_block_load` extrinsics which are not present in the `extrinsics` table.
+pub(crate) async fn missing_extrinsic_blocks(
+	conn: &mut PgConnection,
+	max_block_load: u32,
+) -> Result<Vec<(u32, Vec<u8>, Vec<u8>, u32)>> {
+	let blocks = sqlx::query_as!(
+		BlockExtrinsics,
+		"
+		SELECT block_num, hash, ext, spec FROM blocks
+		WHERE NOT EXISTS
+			(SELECT number FROM extrinsics WHERE extrinsics.number = blocks.block_num)
+		ORDER BY block_num ASC
+		LIMIT $1
+		",
+		i64::from(max_block_load)
+	)
+	.fetch_all(conn)
+	.await?
+	.into_iter()
+	.map(|b| (b.block_num as u32, b.hash, b.ext, b.spec as u32))
+	.collect();
+
+	Ok(blocks)
+}
+
+/// Get upgrade blocks starting from a spec.
+/// Will always return one previous to `from`.
+/// So if you want upgrade specs `from` 30 for polkadot,
+/// this function will also return spec/block_num 29.
+pub(crate) async fn upgrade_blocks_from_spec(conn: &mut sqlx::PgConnection, from: u32) -> Result<HashMap<u32, u32>> {
+	let from = i32::try_from(from)?;
+	let blocks = sqlx::query_as!(
+		BlockNumSpec,
+		r#"
+			SELECT DISTINCT ON (spec) spec, block_num
+			FROM blocks
+			WHERE spec != 0
+			ORDER BY spec, block_num ASC
+		"#,
+	)
+	.fetch_all(conn)
+	.await?
+	.into_iter()
+	.tuple_windows()
+	.filter(|(_curr, next)| next.spec >= from)
```

Comment:
i tried doing this in postgres and 'returning one row previous from the `from` spec, but failed to figure that out. This is simpler/easier way to do it

Reply:
Possibly actually faster too.

Reply:
I had to try this, and the best I have is:

```
WITH 
  exp AS (
    SELECT DISTINCT ON (spec) spec, block_num FROM blocks WHERE spec != 0 ORDER BY spec
  ),
  lagged AS (
    SELECT spec, block_num, LAG(block_num, 1, 0) OVER (ORDER BY spec) as prev_block_num FROM exp
  )
SELECT * FROM lagged WHERE block_num = 314201;
```

I couldn't really understand why, but if I put the `LAG` straight into the `lagged` expr it was all fine until I applied a WHERE clause, and then nothing.... but anyway.. yes, your approach is simpler!

Reply:
Yeah I got about as far, with the same issue -- as soon as `WHERE`is added it filters out the previous spec

Thankfully not too big a deal though